Login Problem

I'm testing and can't log in. But I don't know if I'm doing the right thing. as I read. I must first go to 

`auth/login `

with these parameters 

```
{
  "client_id": "webapp",
  "client_secret": "saqw21!@",
  "username": "super_admin",
  "password": "test123!@#"
}
```

 and that returns this code to me. 

{
  "code": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjbGllbnRJZCI6IndlYmFwcCIsInVzZXJJZCI6MSwiaWF0IjoxNTg2OTU3MzMzLCJleHAiOjE1ODY5NTc1MTMsImF1ZCI6IndlYmFwcCIsImlzcyI6Im1mZ3FmIiwic3ViIjoic3VwZXJfYWRtaW4ifQ.wgHgBirDz1T9YW1pG133XGBwJRM7q61GsgfCRrpC4ws"
}



Then I must go to 

`auth/token` 

with these parameters 

```
{
  "code": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJjbGllbnRJZCI6IndlYmFwcCIsInVzZXJJZCI6MSwiaWF0IjoxNTg2OTU3MzMzLCJleHAiOjE1ODY5NTc1MTMsImF1ZCI6IndlYmFwcCIsImlzcyI6Im1mZ3FmIiwic3ViIjoic3VwZXJfYWRtaW4ifQ.wgHgBirDz1T9YW1pG133XGBwJRM7q61GsgfCRrpC4ws",
  "clientId": "webapp",
  "username": "super_admin"
}
```

so that it gives me the true token but it returns a 401.

```
{
  "error": {
    "statusCode": 401,
    "name": "UnauthorizedError",
    "message": "Invalid Credentials"
  }
}

```
Can someone tell me what I'm doing wrong?
thank in advance @samarpanB

just tested the changes on SFO-7 branch. Doesn't seem like fixing the issue. 👎

Also, the following [line](https://github.com/sourcefuse/loopback4-starter/blob/c7d914050c412a3a26d465cd72f29a85e31f9d8c/src/repositories/user-tenant.repository.ts#L54) throws an error:

> src/repositories/user-tenant.repository.ts:54:9 - error TS2416: Property 'create' in type 'UserTenantRepository' is not assignable to the same property in base type 'DefaultSoftCrudRepository<UserTenant, number | undefined, UserTenantRelations>'.
  Type '(user: User) => Promise<UserTenant>' is not assignable to type '(entity: DataObject<UserTenant>, options?: AnyObject | undefined) => Promise<UserTenant>'.
    Types of parameters 'user' and 'entity' are incompatible.
      Type 'DataObject<UserTenant>' is not assignable to type 'User'.
        Type 'UserTenant' is missing the following properties from type 'User': firstName, lastName, username, defaultTenant, credentials
54   async create(user: User): Promise<UserTenant> {

Got past the same to get going by changing type to `any` instead of `User`.

**Bringing in acceptance tests, I find the login tests all pass. 👍
 Wonder if any issue with boilerplate seed data?**

--- a/.env.defaults
+++ b/.env.defaults
@@ -4,3 +4,4 @@ JWT_SECRET=plmnkoxswqaz
 JWT_ISSUER=lb_api
 USER_TEMP_PASSWORD=temp123!@#
 AUDIT_SCHEMA=logs
+DEFAULT_ROLE=super_admin

--- a/src/__tests__/acceptance/login.controller.acceptance.ts
+++ b/src/__tests__/acceptance/login.controller.acceptance.ts
@@ -1,0 +1,240 @@
+import {Client, expect} from '@loopback/testlab';
+import {Loopback4StarterApplication} from '../..';
+import {setupApplication} from './test-helper';
+import {
+  UserRepository,
+  AuthClientRepository,
+  RoleRepository,
+  TenantRepository,
+  UserTenantRepository,
+} from '../../repositories';
+import {RoleType} from '../../modules/roles/role.enum';
+import {TenantType} from '../../modules/user-tenants/tenant-type.enum';
+
+describe('Login Controller', () => {
+  let app: Loopback4StarterApplication;
+  let client: Client;
+  let userRepo: UserRepository;
+  let authClientRepository: AuthClientRepository;
+  let roleRepository: RoleRepository;
+  let tenantRepository: TenantRepository;
+  let userTenantRepository: UserTenantRepository;
+
+  before('setupApplication', async () => {
+    ({app, client} = await setupApplication());
+  });
+
+  before(givenRepositories);
+  before(setMockData);
+  after(deleteMockData);
+  after(async () => {
+    await app.stop();
+  });
+
+  it('gives status 422 for login request with no client credentials', async () => {
+    const reqData = {};
+    const response = await client
+      .post(`/auth/login`)
+      .send(reqData)
+      .expect(422);
+
+    expect(response).to.have.property('error');
+  });
+
+  it('gives status 422 for login request with no user credentials', async () => {
+    const reqData = {
+      client_id: 'web',
+      client_secret: 'blah',
+    };
+
+    const response = await client
+      .post(`/auth/login`)
+      .send(reqData)
+      .expect(422);
+
+    expect(response).to.have.property('error');
+  });
+
+  it('gives status 401 for login request with wrong client credentials', async () => {
+    const reqData = {
+      client_id: 'web1',
+      client_secret: 'blah1',
+      username: 'super_admin',
+      password: 'test123!@#',
+    };
+
+    const response = await client
+      .post(`/auth/login`)
+      .send(reqData)
+      .expect(401);
+
+    expect(response).to.have.property('error');
+  });
+
+  it('gives status 401 for login request with wrong user credentials', async () => {
+    const reqData = {
+      client_id: 'webapp',
+      client_secret: 'saqw21!@',
+      username: 'someuser',
+      password: 'somepassword',
+    };
+
+    const response = await client
+      .post(`/auth/login`)
+      .send(reqData)
+      .expect(401);
+
+    expect(response).to.have.property('error');
+  });
+
+  it('gives status 200 and return code for login request', async () => {
+    const response = await authLogin();
+    expect(response.status).to.be.equal(200);
+    expect(response.body).to.have.property('code');
+  });
+
+  it('should return refresh token and access token in response', async () => {
+    const authLoginResponse = await authLogin();
+    const authTokenResponse = await authToken(authLoginResponse.body.code);
+
+    expect(authTokenResponse.body).to.have.properties([
+      'accessToken',
+      'refreshToken',
+    ]);
+  });
+
+  it('should return refresh token and access token for token refresh request', async () => {
+    const authLoginResponse = await authLogin();
+    const authTokenResponse = await authToken(authLoginResponse.body.code);
+
+    const response = await client
+      .post(`/auth/token-refresh`)
+      .send({refreshToken: authTokenResponse.body.refreshToken});
+
+    expect(response.body).to.have.properties(['accessToken', 'refreshToken']);
+  });
+
+  it('should return true on logout', async () => {
+    const authLoginResponse = await authLogin();
+    const authTokenResponse = await authToken(authLoginResponse.body.code);
+
+    await client
+      .post(`/logout`)
+      .set('Authorization', `Bearer ${authTokenResponse.body.accessToken}`)
+      .set('refresh-token', `${authTokenResponse.body.refreshToken}`)
+      .expect(200);
+  });
+
+  async function authLogin() {
+    const reqData = {
+      client_id: 'web',
+      client_secret: 'test',
+      username: 'test_user',
+      password: process.env.USER_TEMP_PASSWORD,
+    };
+
+    return await client.post(`/auth/login`).send(reqData);
+  }
+
+  async function authToken(code: string) {
+    const response = await client.post(`/auth/token`).send({
+      clientId: 'web',
+      username: 'test_user',
+      code: code,
+    });
+    return response;
+  }
+
+  async function givenRepositories() {
+    userRepo = await app.getRepository(UserRepository);
+    roleRepository = await app.getRepository(RoleRepository);
+    authClientRepository = await app.getRepository(AuthClientRepository);
+    tenantRepository = await app.getRepository(TenantRepository);
+    userTenantRepository = await app.getRepository(UserTenantRepository);
+  }
+
+  async function deleteMockData() {
+    await userRepo.deleteAllHard();
+    await authClientRepository.deleteAllHard();
+    await userTenantRepository.deleteAll();
+    await roleRepository.deleteAllHard();
+    await tenantRepository.deleteAllHard();
+  }
+
+  async function setMockData() {
+    await roleRepository.createAll([
+      {
+        id: 1,
+        name: 'super_admin',
+        roleKey: RoleType.SuperAdmin,
+        permissions: [
+          'canLoginToIPS',
+          'ViewOwnUser',
+          'ViewAnyUser',
+          'ViewTenantUser',
+          'CreateAnyUser',
+          'CreateTenantUser',
+          'UpdateOwnUser',
+          'UpdateTenantUser',
+          'UpdateAnyUser',
+          'DeleteTenantUser',
+          'DeleteAnyUser',
+          'ViewTenant',
+          'CreateTenant',
+          'UpdateTenant',
+          'DeleteTenant',
+          'ViewRole',
+          'CreateRole',
+          'UpdateRole',
+          'DeleteRole',
+          'ViewAudit',
+          'CreateAudit',
+          'UpdateAudit',
+          'DeleteAudit',
+        ],
+      },
+      {
+        id: 2,
+        name: 'admin',
+        roleKey: RoleType.Admin,
+        permissions: [
+          'ViewOwnUser',
+          'ViewTenantUser',
+          'CreateTenantUser',
+          'UpdateOwnUser',
+          'UpdateTenantUser',
+          'DeleteTenantUser',
+          'ViewTenant',
+          'ViewRole',
+        ],
+      },
+    ]);
+
+    const user = await userRepo.create({
+      id: 1,
+      firstName: 'Test',
+      lastName: 'User',
+      username: 'test_user',
+      defaultTenant: 1,
+    });
+
+    await authClientRepository.create({
+      id: 1,
+      clientId: 'web',
+      clientSecret: 'test',
+      redirectUrl: 'http://localhost:4200/login/success',
+      accessTokenExpiration: 900,
+      refreshTokenExpiration: 86400,
+      authCodeExpiration: 180,
+      secret: 'poiuytrewq',
+    });
+
+    await tenantRepository.create({
+      id: 1,
+      name: 'application',
+      type: TenantType.APPLICATION,
+    });
+
+    await userTenantRepository.create(user);
+  }
+});

--- a/src/__tests__/acceptance/test-helper.ts
+++ b/src/__tests__/acceptance/test-helper.ts
@@ -4,6 +4,7 @@ import {
   givenHttpServerConfig,
   Client,
 } from '@loopback/testlab';
+import {AuthenticationBindings} from 'loopback4-authentication';
 
 export async function setupApplication(): Promise<AppWithClient> {
   const restConfig = givenHttpServerConfig({
@@ -19,6 +20,17 @@ export async function setupApplication(): Promise<AppWithClient> {
   });
 
   await app.boot();
+
+  app.bind('datasources.config.pgdb').to({
+    name: 'pgdb',
+    connector: 'memory',
+  });
+
+  app.bind(AuthenticationBindings.CURRENT_USER).to({
+    id: 1,
+    username: 'test_user',
+    password: process.env.USER_TEMP_PASSWORD,
+  });
   await app.start();
 
   const client = createRestAppClient(app);

--- a/src/__tests__/acceptance/user.controller.acceptance.ts
+++ b/src/__tests__/acceptance/user.controller.acceptance.ts
@@ -1,0 +1,266 @@
+import {Client, expect} from '@loopback/testlab';
+import {Loopback4StarterApplication} from '../..';
+import {setupApplication} from './test-helper';
+import * as jwt from 'jsonwebtoken';
+import {
+  UserRepository,
+  TenantRepository,
+  RoleRepository,
+} from '../../repositories';
+import {TenantType} from '../../modules/user-tenants/tenant-type.enum';
+import {RoleType} from '../../modules/roles/role.enum';
+
+describe('User Controller', () => {
+  let app: Loopback4StarterApplication;
+  let client: Client;
+  let userRepo: UserRepository;
+  let tenantRepository: TenantRepository;
+  let roleRepository: RoleRepository;
+
+  before('setupApplication', async () => {
+    ({app, client} = await setupApplication());
+  });
+
+  before(givenRepositories);
+
+  afterEach(deleteMockData);
+  after(async () => {
+    await app.stop();
+  });
+
+  it('gives status 422 when data sent is incorrect', async () => {
+    const reqData = {};
+    const response = await client
+      .post(`/users`)
+      .send(reqData)
+      .expect(422);
+
+    expect(response).to.have.property('error');
+  });
+
+  it('gives status 401 when no token is passed', async () => {
+    const response = await client.get(`/users`).expect(401);
+
+    expect(response).to.have.property('error');
+  });
+
+  it('gives status 200 when token is passed', async () => {
+    const testUser = {
+      id: 1,
+      username: 'test_user',
+      password: 'test_password',
+      permissions: [
+        'CreateAnyUser',
+        'UpdateAnyUser',
+        'ViewAnyUser',
+        'ViewAnyUser',
+      ],
+    };
+    const token = jwt.sign(testUser, process.env.JWT_SECRET as string, {
+      expiresIn: 180000,
+      issuer: process.env.JWT_ISSUER,
+    });
+
+    await client
+      .get(`/users`)
+      .set('authorization', `Bearer ${token}`)
+      .expect(200);
+  });
+
+  it('gives status 422 when user details are not correct', async () => {
+    const testUser = {
+      id: 1,
+      username: 'test_user',
+      password: 'test_password',
+      permissions: [
+        'CreateAnyUser',
+        'UpdateAnyUser',
+        'ViewAnyUser',
+        'DeleteAnyUser',
+      ],
+    };
+    const token = jwt.sign(testUser, process.env.JWT_SECRET as string, {
+      expiresIn: 180000,
+      issuer: process.env.JWT_ISSUER,
+    });
+
+    const userToAdd = {};
+
+    await client
+      .post(`/users`)
+      .set('authorization', `Bearer ${token}`)
+      .send(userToAdd)
+      .expect(422);
+  });
+
+  it('gives status 200 and user details when user is added', async () => {
+    const token = await getToken();
+    const reqToAddUser = await postUser(token);
+
+    expect(reqToAddUser.status).to.be.equal(200);
+    const response = await client
+      .get(`/users/${reqToAddUser.body.id}`)
+      .set('authorization', `Bearer ${token}`)
+      .expect(200);
+
+    expect(response.body).to.have.properties([
+      'username',
+      'phone',
+      'firstName',
+    ]);
+    expect(response.body.username).to.be.equal('test_number');
+  });
+
+  it('updates user successfully', async () => {
+    const token = await getToken();
+    const reqToAddUser = await postUser(token);
+
+    const userToUpdate = {
+      firstName: 'New Test',
+      lastName: 'User',
+      username: 'number',
+      phone: '9898989898',
+      defaultTenant: 1,
+    };
+
+    await client
+      .put(`/users/${reqToAddUser.body.id}`)
+      .set('authorization', `Bearer ${token}`)
+      .send(userToUpdate)
+      .expect(204);
+
+    const response = await client
+      .get(`/users/${reqToAddUser.body.id}`)
+      .set('authorization', `Bearer ${token}`)
+      .expect(200);
+
+    expect(response.body).to.have.properties([
+      'username',
+      'phone',
+      'firstName',
+    ]);
+    expect(response.body.username).to.be.equal('number');
+  });
+
+  it('deletes a user successfully', async () => {
+    const token = await getToken();
+    const reqToAddUser = await postUser(token);
+
+    await client
+      .del(`/users/${reqToAddUser.body.id}`)
+      .set('authorization', `Bearer ${token}`)
+      .expect(204);
+  });
+
+  it('should return count', async () => {
+    const token = await getToken();
+
+    await client
+      .get(`/users/count`)
+      .set('authorization', `Bearer ${token}`)
+      .expect(200);
+  });
+
+  async function getToken() {
+    const testUser = {
+      id: 1,
+      username: 'test_user',
+      password: 'test_password',
+      permissions: [
+        'CreateAnyUser',
+        'UpdateAnyUser',
+        'ViewAnyUser',
+        'DeleteAnyUser',
+      ],
+    };
+    const token = jwt.sign(testUser, process.env.JWT_SECRET as string, {
+      expiresIn: 180000,
+      issuer: process.env.JWT_ISSUER,
+    });
+    return token;
+  }
+
+  async function postUser(token: string) {
+    await roleRepository.createAll([
+      {
+        id: 1,
+        name: 'super_admin',
+        roleKey: RoleType.SuperAdmin,
+        permissions: [
+          'canLoginToIPS',
+          'ViewOwnUser',
+          'ViewAnyUser',
+          'ViewTenantUser',
+          'CreateAnyUser',
+          'CreateTenantUser',
+          'UpdateOwnUser',
+          'UpdateTenantUser',
+          'UpdateAnyUser',
+          'DeleteTenantUser',
+          'DeleteAnyUser',
+          'ViewTenant',
+          'CreateTenant',
+          'UpdateTenant',
+          'DeleteTenant',
+          'ViewRole',
+          'CreateRole',
+          'UpdateRole',
+          'DeleteRole',
+          'ViewAudit',
+          'CreateAudit',
+          'UpdateAudit',
+          'DeleteAudit',
+        ],
+      },
+      {
+        id: 2,
+        name: 'admin',
+        roleKey: RoleType.Admin,
+        permissions: [
+          'ViewOwnUser',
+          'ViewTenantUser',
+          'CreateTenantUser',
+          'UpdateOwnUser',
+          'UpdateTenantUser',
+          'DeleteTenantUser',
+          'ViewTenant',
+          'ViewRole',
+        ],
+      },
+    ]);
+
+    await tenantRepository.create({
+      id: 1,
+      name: 'application',
+      type: TenantType.APPLICATION,
+    });
+
+    const userToAdd = {
+      id: 1,
+      firstName: 'Test',
+      lastName: 'User',
+      username: 'test_number',
+      phone: '9898989898',
+      defaultTenant: 1,
+    };
+
+    const reqToAddUser = await client
+      .post(`/users`)
+      .set('authorization', `Bearer ${token}`)
+      .send(userToAdd);
+
+    return reqToAddUser;
+  }
+
+  async function deleteMockData() {
+    await userRepo.deleteAllHard();
+    await roleRepository.deleteAllHard();
+    await tenantRepository.deleteAllHard();
+  }
+
+  async function givenRepositories() {
+    userRepo = await app.getRepository(UserRepository);
+    roleRepository = await app.getRepository(RoleRepository);
+    tenantRepository = await app.getRepository(TenantRepository);
+  }
+});

--- a/src/modules/user-tenants/user.controller.ts
+++ b/src/modules/user-tenants/user.controller.ts
@@ -15,18 +15,21 @@ import {
   post,
   put,
   requestBody,
+  HttpErrors,
 } from '@loopback/rest';
 import {authenticate, STRATEGY} from 'loopback4-authentication';
 import {authorize} from 'loopback4-authorization';
 
 import {User} from '../../models';
-import {UserRepository} from '../../repositories';
+import {UserRepository, UserTenantRepository} from '../../repositories';
 import {PermissionKey} from '../auth/permission-key.enum';
 
 export class UserController {
   constructor(
     @repository(UserRepository)
     public userRepository: UserRepository,
+    @repository(UserTenantRepository)
+    public userTenantRepo: UserTenantRepository,
   ) {}
 
   @authenticate(STRATEGY.BEARER)
@@ -40,7 +43,15 @@ export class UserController {
     },
   })
   async create(@requestBody() user: User): Promise<User> {
-    return await this.userRepository.create(user);
+    if (!user.id || !user.defaultTenant) {
+      throw new HttpErrors.UnprocessableEntity(
+        'User Id or Default Tenant Id is missing in the request parameters',
+      );
+    }
+    const response = await this.userRepository.create(user);
+    await this.userTenantRepo.create(user);
+
+    return response;
   }
 
   @authenticate(STRATEGY.BEARER)

--- a/src/repositories/user-tenant.repository.ts
+++ b/src/repositories/user-tenant.repository.ts
@@ -7,6 +7,7 @@ import {DefaultSoftCrudRepository} from './default-soft-crud.repository.base';
 import {RoleRepository} from './role.repository';
 import {TenantRepository} from './tenant.repository';
 import {UserRepository} from './user.repository';
+import {HttpErrors} from '@loopback/rest';
 
 export class UserTenantRepository extends DefaultSoftCrudRepository<
   UserTenant,
@@ -29,6 +30,8 @@ export class UserTenantRepository extends DefaultSoftCrudRepository<
     userRepositoryGetter: Getter<UserRepository>,
     @repository.getter(RoleRepository)
     roleRepositoryGetter: Getter<RoleRepository>,
+    @repository(RoleRepository)
+    public roleRepo: RoleRepository,
   ) {
     super(UserTenant, dataSource);
 
@@ -46,5 +49,27 @@ export class UserTenantRepository extends DefaultSoftCrudRepository<
       'role_id',
       roleRepositoryGetter,
     );
+  }
+
+  async create(user: User): Promise<UserTenant> {
+    if (!user.id || !user.defaultTenant) {
+      throw new HttpErrors.UnprocessableEntity(
+        'User Id or Default Tenant Id is missing in the request parameters',
+      );
+    }
+    const userTenant: UserTenant = new UserTenant();
+    userTenant.userId = user.id;
+    userTenant.tenantId = user.defaultTenant;
+    const role = await this.roleRepo.findOne({
+      where: {
+        name: process.env.DEFAULT_ROLE,
+      },
+    });
+    if (role && role.id) {
+      userTenant.roleId = role.id;
+    } else {
+      throw new HttpErrors.InternalServerError('Failed to set default role.');
+    }
+    return await super.create(userTenant);
   }
 }

--- a/src/repositories/user-tenant.repository.ts
+++ b/src/repositories/user-tenant.repository.ts
@@ -68,7 +68,7 @@ export class UserTenantRepository extends DefaultSoftCrudRepository<
     if (role && role.id) {
       userTenant.roleId = role.id;
     } else {
-      throw new HttpErrors.InternalServerError('Failed to set default role.');
+      throw new HttpErrors.UnprocessableEntity('Failed to set default role.');
     }
     return await super.create(userTenant);
   }

